lesson-4-metrics-logs-observability/README.md

```diff
@@ -0,0 +1,96 @@
+Microservices Perf\&Scale testing
+=================================
+
+What is handy to know:
+
+* What is percentile (and histogram and average)
+* Sequential vs. parallel
+* Throughput vs. latency
+* Horizontal vs. vertical scaling
+* Understand pod resource **requests** and **limits** (quality of service for testng vs. prod settings)
+
+![Percentiles](assets/percentile.svg "Percentiles")
+
+Preparing for testing:
+
+* Do we have knowledge and/or some diagrams on how the application works and how it interacts with the rest of the ecosystem
+* How to deploy the application into non-production environment for the testing
+* What (perf and scale related) risks the application/deployment have
+* What scenario would be testable from a performance point of view
+  * What if you have to pick one, which one?
+* Any existing perf&scale results
+* What are the perf&scale expectations/projections
+
+Example application
+-------------------
+
+How it works:
+
+![Example application - DB schema](assets/example-app-db-schema.svg "Example application - DB schema")
+
+Name: Banking-like™ application
+
+* Manage users and watch their money balance
+* Add money transfer transactions
+* Search users by name, address, email…
+
+Technicalities:
+
+* PostgreSQL with just two tables `users` and `moves`
+* Python backend application in Flask framework with SQLAlchemy as an ORM
+* Gunicorn as an application server
+
+Source code: <https://github.com/jhutar/perfscale-demo-app/blob/main/myapp.py>
+
+Deployment:
+
+![Example application - deployment](assets/example-app-deployment.svg "Example application - deployment")
+
+Source code: <https://github.com/jhutar/perfscale-demo-app/blob/main/deploy.yaml>
+
+Risks:
+
+* How many API requests it can handle
+* Is there a difference in read/write operations
+* Will performance drop with more entities in the DB
+* What resources are we going to need to run the application
+* What aspects of the application to monitor
+* Can UI handle the traffic
+
+Scenario:
+
+We will stick to just one question: concurrent API read access
+
+Locust script to the rescue: <https://github.com/jhutar/perfscale-demo-app/blob/main/testing.py>
+
+Existing results:
+
+None :-)
+
+Expectations:
+
+* In peaks 100 users concurrently
+* Plan to grow to 1000 concurrent users next year, double it year after…
+* Every user does 3 transactions every day
+* Expected latency 99.9th percentile below 100 ms
+
+Demo?
+-----
+
+Would scaling the application horizontally improve RPS?
+
+Can we scale indefinetelly?
+
```

Comment:
Could you please improve this section with actual commands that you are planning to run during the demo so people can easily copy/paste them? I can see that your app has a deploy.yaml

Reply:
Hello! Added stuff into the section. What do you think?